crypto/property/property.c

```diff
@@ -461,33 +463,39 @@ static void alg_do_one(ALGORITHM *alg, IMPLEMENTATION *impl,
     fn(alg->nid, impl->method.method, fnarg);
 }
 
-struct alg_do_each_data_st {
-    void (*fn)(int id, void *method, void *fnarg);
-    void *fnarg;
-};
-
-static void alg_do_each(ossl_uintmax_t idx, ALGORITHM *alg, void *arg)
+static void alg_copy(ossl_uintmax_t idx, ALGORITHM *alg, void *arg)
 {
-    struct alg_do_each_data_st *data = arg;
-    int i, end = sk_IMPLEMENTATION_num(alg->impls);
+    STACK_OF(ALGORITHM) *newalg = arg;
 
-    for (i = 0; i < end; i++) {
-        IMPLEMENTATION *impl = sk_IMPLEMENTATION_value(alg->impls, i);
-
-        alg_do_one(alg, impl, data->fn, data->fnarg);
-    }
+    (void)sk_ALGORITHM_push(newalg, alg);
 }
 
 void ossl_method_store_do_all(OSSL_METHOD_STORE *store,
                               void (*fn)(int id, void *method, void *fnarg),
                               void *fnarg)
 {
-    struct alg_do_each_data_st data;
+    int i, j;
+    int numalgs, numimps;
+    STACK_OF(ALGORITHM) *tmpalgs = sk_ALGORITHM_new_null();
```

Comment:
Ditch the creation here and do:
`tmpalgs = OPENSSL_sk_new_reserve(NULL, ossl_sa_ALGORITHM_num(store->algs));`
under the read lock.  This saves reallocation of the stack.

Alternatively, keep this and call `OPENSSL_sk_reserve(ossl_sa_ALGORITHM_num(store->algs));` inside the read lock.  This is what happens internally in _new_reserve, so no performance difference.

Both approaches would allow the failure check for stack insertions to be skipped, although it's prudent to retain them.